fuse_core/src/timestamp_manager.cpp

```diff
@@ -100,9 +100,26 @@ void TimestampManager::query(
       const rclcpp::Time & current_stamp = *current_iter;
       // Check if the timestamp pair is exactly an existing pair. If so, don't add it.
       auto history_iter = motion_model_history_.lower_bound(previous_stamp);
+      rclcpp::Time history_beginning_stamp = history_iter->second.beginning_stamp;
+      rclcpp::Time history_ending_stamp = history_iter->second.ending_stamp;
+
+      // lower_bound() can default construct Time, which causes a mismatched clock error
```

Comment:
This explicit check was the last thing I added (after adding RCL_ROS_TIME to all default constructed time objects that I could find.)

I had assumed that lower_bound() would end up returning default constructed instances instead of end()... But since it doesn't it means that there looks to be an initial default constructed(??) stamp getting added somewhere (but I don't see a place where it could get inserted :scream: )

I think it'll take me awhile to trace, so I might TODO this since the code works as of now.
I'll update the comment accordingly.

---

Also I think this is unrelated but I am aware of a single comparison somewhere between two instances of rclcpp::Time(0, 0, **RCL_CLOCK_UNINITIALIZED**) that happens on init, but I couldn't trace where they were coming from.

Reply:
Comment updated

Reply:
I did some debugging with where the mismatched timestamps were coming from. It looks like it was the construction of the `TimestampManager::MotionModelSegment()` here: https://github.com/locusrobotics/fuse/pull/307/files#diff-30ff4b063bdf335221719ed9125d42c7473a269f56d2a638392bf8f482e7a723R154

That class had two time stamps `begginning_stamp` and `ending_stamp` that were being default constructed. I fixed it by removing the default constructor and replacing it with one that takes a clock type and uses it to initialize those stamps. The advantage is it allows `TimestampManager` to use any clock type, as long as the same clock is consistently used. I made sure the `test_timestamp_manager.cpp` test worked with the `RCL_ROS_TIME` insertions, then reverted them and found it still works :tada: